README.md
# PSV Shell Plus

This is a fork of PSVshell, called PSVShell+. It adds a number of extra features,  mostly for PS TV owners
(like myself), but others can find it useful too.

## Extra features

- Swap the cross and cirle buttons for playing japanese games
- Disable **L3** and **R3** (the stick buttons) to prevent you from accidentally pressing them
- DS4 touchpad emulation. This has three modes:
  - **front:** Use the thouchpad to emulate the front touchscreen
  - **F/B:** Emulated both the front and the back thouchpad, usign the touchpad button to toggle between them
  - **F/B/X:** Same as "F/B", but it toggles between front, back and off, not just front and back
- DS4 motion emulation. This has two modes:
  - **normal:** The DS4 controller has the same orientation as the Vita
  - **vertical:** The DS4 held horizontally has the orientation of the Vita held vertically
- Reboot or shutdown the Vita. This just seemed nice to have on hand.

In the "FULL" mode, there are now two pages, that you can switch between using **L** and **R**. The main page is the
same as it was in PSVshell. All the extra features added by PSVShell+ are on the second page.

Saving the profile on the main page (below the clock speeds) also save the extra features in the same manner.

## Acknowledgments

- To [Electry](https://github.com/Electry) for the original [PSVshell](https://github.com/Electry/PSVshell)
- To [MERLev](https://github.com/MERLev) for [ds4Touch](https://github.com/MERLev/ds4Touch)
- To [OperationNT414C](https://github.com/OperationNT414C) for [DSMotion](https://github.com/OperationNT414C/DSMotion)
- And a big thank you to [VitaSDK](https://github.com/vitasdk) for making this all possible.

# Original readme
Yet another overclocking plugin

## Features:
- Allows you to change CPU, GPU, BUS and XBAR clocks separately, in these steps:
  - **CPU:** 41, 83, 111, 166, 222, 333, 444, 500 MHz
  - **GPU (ES4):** 41, 55, 83, 111, 166, 222 MHz
  - **BUS:** 55, 83, 111, 166, 222 MHz
  - **XBAR:** 83, 111, 166 MHz
- Supports per-app profiles
- Shows per-core CPU usage in %, including peak single-thread load
- Runs in kernelland (=> visible in LiveArea)
- Pretty GUI with some useless eye-candy metrics such as ram/vram usage, battery temp, etc...
- Does not slow down games when menu is open
- Does not crash Adrenaline
- Clean code and patches
- **3.60** and **3.65** FWs are supported

## How to use:
- Press **SELECT + UP** or **SELECT + DOWN** to toggle between 3 GUI modes

#### When in 'FULL' mode:
- Use **UP/DOWN** to move in the menu
- Press **X** to toggle frequency mode for currently selected **> device <**:
  - **Default freq.** (WHITE) - the plugin will not interfere, but rather use the default freq. for current game
  - **Manual freq.** (BLUE) - the plugin will use your specified freq.
    - press **LEFT/RIGHT** to immediately change the frequency
- Press **X** when **> save profile <** is selected to save/delete profiles
  - All **Manual freq.** (BLUE) will be loaded and applied next time you start/resume the game
  - All **Default freq.** (WHITE) will be kept to default (set to whatever freq. the game asks for)
- Press and hold **LEFT TRIGGER** and **> save profile <** will change to **> save global <**
  - Press **X** when **> save global <** is selected and the options will be saved to *global* (default) profile
  - *Global* profile will be used as default profile when game-specific profile doesn't exist

## Screenshots:
![2019-12-21-181613](https://user-images.githubusercontent.com/12598379/71311342-c15df300-241e-11ea-8baf-c67ec2bcbbd7.png)

### 'FPS only' mode
![2019-11-01-194930](https://user-images.githubusercontent.com/12598379/68051962-e10f4f80-fce8-11e9-92d0-9662cc6f0d04.png)

### 'HUD' mode
![2019-12-21-181809](https://user-images.githubusercontent.com/12598379/71311344-c1f68980-241e-11ea-9ca1-4207d4887002.png)

### 'FULL' mode
![2019-12-21-181801](https://user-images.githubusercontent.com/12598379/71311343-c1f68980-241e-11ea-8249-5f2e0c44d642.png)

## Credits:
- [Yifan Lu](https://github.com/yifanlu) - for [ScePervasive](https://wiki.henkaku.xyz/vita/Pervasive) RE
- [dots-tb](https://github.com/dots-tb) - for ksceKernelInvokeProcEventHandler() hook
- [Rinnegatamante](https://github.com/Rinnegatamante) - for orig. framecounter impl.


Changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to "Ad Hoc" versioning (meaning, I assign versions based on how I like it).

## [Unreleased]
### Added
- FPS Limiter, that's active when only the FPS is visible.
- Page indicator dots to the top of the widget in full mode.
- Changelog

### Changed
- Updated code for new VitaSDK version (2022-03-10).
- The FPS counter is now red, when the FPS limiter is enabled.

## [1.0] - 2021-08-11
### Added
- Paging to the widget in full mode, with the L and R buttons.
- A second page for additional settings.
- Option to swap the X and O buttons.
- Option to disable the L3 and L3 buttons.
- Support for touch controls on DS4 controllers via Bluetooth.
- Support for motion controls on DS4 controllers via Bluetooth.

### Changed
- Updated the README.md with the additional features

[Unreleased]: https://onedev.server.local/projects/2/blob
[1.0]: https://onedev.server.local/projects/2/blob/1.0

src/oc.c
#include "headers.h"

// Declare helper getter/setter for GpuEs4
static int __kscePowerGetGpuEs4ClockFrequency() {
    int a1, a2;
    _kscePowerGetGpuEs4ClockFrequency(&a1, &a2);
    return a1;
}
static int __kscePowerSetGpuEs4ClockFrequency(int freq) {
    return _kscePowerSetGpuEs4ClockFrequency(freq, freq);
}

// Declare static getters/setters
PSVS_OC_DECL_GETTER(_kscePowerGetArmClockFrequency);
PSVS_OC_DECL_GETTER(_kscePowerGetBusClockFrequency);
PSVS_OC_DECL_GETTER(_kscePowerGetGpuXbarClockFrequency);
PSVS_OC_DECL_SETTER(_kscePowerSetArmClockFrequency);
PSVS_OC_DECL_SETTER(_kscePowerSetBusClockFrequency);
PSVS_OC_DECL_SETTER(_kscePowerSetGpuXbarClockFrequency);

static psvs_oc_devopt_t g_oc_devopt[PSVS_OC_DEVICE_MAX] = {
    [PSVS_OC_DEVICE_CPU] = {
        .freq_n = 8, .freq = {41, 83, 111, 166, 222, 333, 444, 500}, .default_freq = 333,
        .get_freq = __kscePowerGetArmClockFrequency,
        .set_freq = __kscePowerSetArmClockFrequency
    },
    [PSVS_OC_DEVICE_GPU_ES4] = {
        .freq_n = 6, .freq = {41, 55, 83, 111, 166, 222}, .default_freq = 111,
        .get_freq = __kscePowerGetGpuEs4ClockFrequency,
        .set_freq = __kscePowerSetGpuEs4ClockFrequency
    },
    [PSVS_OC_DEVICE_BUS] = {
        .freq_n = 5, .freq = {55, 83, 111, 166, 222}, .default_freq = 222,
        .get_freq = __kscePowerGetBusClockFrequency,
        .set_freq = __kscePowerSetBusClockFrequency
    },
    [PSVS_OC_DEVICE_GPU_XBAR] = {
        .freq_n = 3, .freq = {83, 111, 166}, .default_freq = 111,
        .get_freq = __kscePowerGetGpuXbarClockFrequency,
        .set_freq = __kscePowerSetGpuXbarClockFrequency
    },
};

int psvs_oc_get_freq(psvs_oc_device_t device) {
    return g_oc_devopt[device].get_freq();
}

int psvs_oc_set_freq(psvs_oc_device_t device, int freq) {
    return g_oc_devopt[device].set_freq(freq);
}

void psvs_oc_holy_shit() {
    // Apply mul:div (15:0)
    ScePervasiveForDriver_0xE9D95643(15, 16 - 0);

    // Store global freq & mul for kscePowerGetArmClockFrequency()
    *ScePower_41C8 = 500;
    *ScePower_41CC = 15;
}

int psvs_oc_get_target_freq(psvs_oc_device_t device, int default_freq) {
    if (g_profile.mode[device] == PSVS_OC_MODE_MANUAL)
        return g_profile.manual_freq[device];
    return default_freq;
}

void psvs_oc_set_target_freq(psvs_oc_device_t device) {
    // Refresh manual clocks
    if (g_profile.mode[device] == PSVS_OC_MODE_MANUAL)
        psvs_oc_set_freq(device, g_profile.manual_freq[device]);
    // Restore default clocks
    else if (g_profile.mode[device] == PSVS_OC_MODE_DEFAULT)
        psvs_oc_set_freq(device, psvs_oc_get_default_freq(device));
}

psvs_oc_mode_t psvs_oc_get_mode(psvs_oc_device_t device) {
    return g_profile.mode[device];
}

void psvs_oc_set_mode(psvs_oc_device_t device, psvs_oc_mode_t mode) {
    g_profile.mode[device] = mode;
    g_profile_has_changed = true;
    psvs_oc_set_target_freq(device);
}

int psvs_oc_get_default_freq(psvs_oc_device_t device) {
    int freq = g_oc_devopt[device].default_freq;

    if (g_pid == INVALID_PID)
        return freq;

    uintptr_t pstorage = 0;
    int ret = ksceKernelGetProcessLocalStorageAddrForPid(g_pid, *ScePower_0, (void **)&pstorage, 0);
    if (ret < 0 || pstorage == 0)
        return freq;

    switch (device) {
        case PSVS_OC_DEVICE_BUS: freq = *(uint32_t *)(pstorage + 40); break;
        default:
        case PSVS_OC_DEVICE_CPU: freq = *(uint32_t *)(pstorage + 44); break;
        case PSVS_OC_DEVICE_GPU_XBAR: freq = *(uint32_t *)(pstorage + 48); break;
        case PSVS_OC_DEVICE_GPU_ES4: freq = *(uint32_t *)(pstorage + 52); break;
    }

    // Validate
    bool valid = false;
    for (int i = 0; i < g_oc_devopt[device].freq_n; i++) {
        if (freq == g_oc_devopt[device].freq[i]) {
            valid = true;
            break;
        }
    }

    return valid ? freq : g_oc_devopt[device].default_freq;
}

void psvs_oc_reset_manual(psvs_oc_device_t device) {
    g_profile.manual_freq[device] = psvs_oc_get_freq(device);
    g_profile_has_changed = true;
}

void psvs_oc_change_manual(psvs_oc_device_t device, bool raise_freq) {
    int target_freq = g_profile.manual_freq[device]; // current manual freq

    for (int i = 0; i < g_oc_devopt[device].freq_n; i++) {
        int ii = raise_freq ? i : g_oc_devopt[device].freq_n - i - 1;
        if ((raise_freq && g_oc_devopt[device].freq[ii] > target_freq)
                || (!raise_freq && g_oc_devopt[device].freq[ii] < target_freq)) {
            target_freq = g_oc_devopt[device].freq[ii];
            break;
        }
    }

    g_profile.manual_freq[device] = target_freq;
    g_profile_has_changed = true;

    // Refresh manual clocks
    if (g_profile.mode[device] == PSVS_OC_MODE_MANUAL)
        psvs_oc_set_freq(device, g_profile.manual_freq[device]);
}

void psvs_oc_init() {
    g_profile_has_changed = true;
    for (int i = 0; i < PSVS_OC_DEVICE_MAX; i++) {
        g_profile.mode[i] = PSVS_OC_MODE_DEFAULT;
        psvs_oc_reset_manual(i);
    }
}
